pkg/apis/experimental/types.go

```diff
@@ -462,3 +462,79 @@ type JobCondition struct {
 	// Human readable message indicating details about last transition.
 	Message string `json:"message,omitempty"`
 }
+
+// An Ingress is a way to give services externally-reachable urls. Each Ingress is a
+// collection of rules that allow inbound connections to reach the endpoints defined by
+// a backend.
+type Ingress struct {
+	unversioned.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	api.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the desired state of the Ingress.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
+	Spec IngressSpec `json:"spec,omitempty"`
+
+	// Status is the current state of the Ingress.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
+	Status IngressStatus `json:"status,omitempty"`
+}
+
+// IngressList is a collection of Ingress.
+type IngressList struct {
+	unversioned.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	unversioned.ListMeta `json:"metadata,omitempty"`
+
+	// Items is the list of Ingress.
+	Items []Ingress `json:"items"`
+}
+
+// IngressSpec describes the Ingress the user wishes to exist.
+type IngressSpec struct {
+	// A list of rules used to configure the Ingress.
+	// http://<host>:<port>/<path>?<searchpart> -> IngressBackend
+	// Where parts of the url conform to RFC 1738.
+	Rules []IngressRule `json:"rules,omitempty"`
```

Comment:
Rules are not omitempty if one specifies a spec (to me this is just like PodSpec and Containers, podspec is omittable but containers are not).